In [1]:
import pandas as pd
import numpy as np
from numpy import average
import matplotlib.pyplot as plt 
import matplotlib.mlab as mlab
import matplotlib.patches as mpatches
import seaborn as sb
import datetime as dt
import glob
import json
from collections import Counter

# called to counteract dark-mode Jupyter from swallowing axes &c.
plt.style.use('seaborn-whitegrid')

In [2]:
# Source citation: Fanaee-T, Hadi, and Gama, Joao, 'Event labeling combining ensemble detectors and background knowledge', 
# Progress in Artificial Intelligence (2013): pp. 1-15, Springer Berlin Heidelberg (http://dx.doi.org/10.1007/s13748-013-0040-3) 
# (accessed: 15 Mar 2022)

# Import the core datasets
daily_core = pd.read_csv('data/uci_data/day.csv')
hourly_core = pd.read_csv('data/uci_data/hour.csv')

<hr>

A fly-by view of the data. Looking for things that poke out.

In [3]:
daily_core.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

In [4]:
hourly_core.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

In [5]:
daily_core.head().T

0           1           2           3           4
instant              1           2           3           4           5
dteday      2011-01-01  2011-01-02  2011-01-03  2011-01-04  2011-01-05
season               1           1           1           1           1
yr                   0           0           0           0           0
mnth                 1           1           1           1           1
holiday              0           0           0           0           0
weekday              6           0           1           2           3
workingday           0           0           1           1           1
weathersit           2           2           1           1           1
temp          0.344167    0.363478    0.196364         0.2    0.226957
atemp         0.363625    0.353739    0.189405    0.212122     0.22927
hum           0.805833    0.696087    0.437273    0.590435    0.436957
windspeed     0.160446    0.248539    0.248309    0.160296      0.1869
casual             331         131         120         108          82
registered         654         670        1229        1454        1518
cnt                985         801        1349        1562        1600

In [6]:
hourly_core.head().T

0           1           2           3           4
instant              1           2           3           4           5
dteday      2011-01-01  2011-01-01  2011-01-01  2011-01-01  2011-01-01
season               1           1           1           1           1
yr                   0           0           0           0           0
mnth                 1           1           1           1           1
hr                   0           1           2           3           4
holiday              0           0           0           0           0
weekday              6           6           6           6           6
workingday           0           0           0           0           0
weathersit           1           1           1           1           1
temp              0.24        0.22        0.22        0.24        0.24
atemp           0.2879      0.2727      0.2727      0.2879      0.2879
hum               0.81         0.8         0.8        0.75        0.75
windspeed          0.0         0.0         0.0         0.0         0.0
casual               3           8           5           3           0
registered          13          32          27          10           1
cnt                 16          40          32          13           1

In [7]:
daily_core.shape

(731, 16)

In [8]:
hourly_core.shape

(17379, 17)

In [9]:
daily_core.describe().T

count         mean          std        min          25%  \
instant     731.0   366.000000   211.165812   1.000000   183.500000   
season      731.0     2.496580     1.110807   1.000000     2.000000   
yr          731.0     0.500684     0.500342   0.000000     0.000000   
mnth        731.0     6.519836     3.451913   1.000000     4.000000   
holiday     731.0     0.028728     0.167155   0.000000     0.000000   
weekday     731.0     2.997264     2.004787   0.000000     1.000000   
workingday  731.0     0.683995     0.465233   0.000000     0.000000   
weathersit  731.0     1.395349     0.544894   1.000000     1.000000   
temp        731.0     0.495385     0.183051   0.059130     0.337083   
atemp       731.0     0.474354     0.162961   0.079070     0.337842   
hum         731.0     0.627894     0.142429   0.000000     0.520000   
windspeed   731.0     0.190486     0.077498   0.022392     0.134950   
casual      731.0   848.176471   686.622488   2.000000   315.500000   
registered  731.0  3656.172367  1560.256377  20.000000  2497.000000   
cnt         731.0  4504.348837  1937.211452  22.000000  3152.000000   

                    50%          75%          max  
instant      366.000000   548.500000   731.000000  
season         3.000000     3.000000     4.000000  
yr             1.000000     1.000000     1.000000  
mnth           7.000000    10.000000    12.000000  
holiday        0.000000     0.000000     1.000000  
weekday        3.000000     5.000000     6.000000  
workingday     1.000000     1.000000     1.000000  
weathersit     1.000000     2.000000     3.000000  
temp           0.498333     0.655417     0.861667  
atemp          0.486733     0.608602     0.840896  
hum            0.626667     0.730209     0.972500  
windspeed      0.180975     0.233214     0.507463  
casual       713.000000  1096.000000  3410.000000  
registered  3662.000000  4776.500000  6946.000000  
cnt         4548.000000  5956.000000  8714.000000

In [10]:
hourly_core.describe().T

count         mean          std   min        25%        50%  \
instant     17379.0  8690.000000  5017.029500  1.00  4345.5000  8690.0000   
season      17379.0     2.501640     1.106918  1.00     2.0000     3.0000   
yr          17379.0     0.502561     0.500008  0.00     0.0000     1.0000   
mnth        17379.0     6.537775     3.438776  1.00     4.0000     7.0000   
hr          17379.0    11.546752     6.914405  0.00     6.0000    12.0000   
holiday     17379.0     0.028770     0.167165  0.00     0.0000     0.0000   
weekday     17379.0     3.003683     2.005771  0.00     1.0000     3.0000   
workingday  17379.0     0.682721     0.465431  0.00     0.0000     1.0000   
weathersit  17379.0     1.425283     0.639357  1.00     1.0000     1.0000   
temp        17379.0     0.496987     0.192556  0.02     0.3400     0.5000   
atemp       17379.0     0.475775     0.171850  0.00     0.3333     0.4848   
hum         17379.0     0.627229     0.192930  0.00     0.4800     0.6300   
windspeed   17379.0     0.190098     0.122340  0.00     0.1045     0.1940   
casual      17379.0    35.676218    49.305030  0.00     4.0000    17.0000   
registered  17379.0   153.786869   151.357286  0.00    34.0000   115.0000   
cnt         17379.0   189.463088   181.387599  1.00    40.0000   142.0000   

                   75%         max  
instant     13034.5000  17379.0000  
season          3.0000      4.0000  
yr              1.0000      1.0000  
mnth           10.0000     12.0000  
hr             18.0000     23.0000  
holiday         0.0000      1.0000  
weekday         5.0000      6.0000  
workingday      1.0000      1.0000  
weathersit      2.0000      4.0000  
temp            0.6600      1.0000  
atemp           0.6212      1.0000  
hum             0.7800      1.0000  
windspeed       0.2537      0.8507  
casual         48.0000    367.0000  
registered    220.0000    886.0000  
cnt           281.0000    977.0000

In [11]:
daily_core.nunique(axis=0)

instant       731
dteday        731
season          4
yr              2
mnth           12
holiday         2
weekday         7
workingday      2
weathersit      3
temp          499
atemp         690
hum           595
windspeed     650
casual        606
registered    679
cnt           696
dtype: int64

In [12]:
hourly_core.nunique(axis=0)

instant       17379
dteday          731
season            4
yr                2
mnth             12
hr               24
holiday           2
weekday           7
workingday        2
weathersit        4
temp             50
atemp            65
hum              89
windspeed        30
casual          322
registered      776
cnt             869
dtype: int64

Whoops, what's up with *weathersit*? There is one missing in the daily set.

In [13]:
print(daily_core['weathersit'].unique())
print(hourly_core['weathersit'].unique())

[2 1 3]
[1 2 3 4]


In [14]:
daily_core.dtypes

instant         int64
dteday         object
season          int64
yr              int64
mnth            int64
holiday         int64
weekday         int64
workingday      int64
weathersit      int64
temp          float64
atemp         float64
hum           float64
windspeed     float64
casual          int64
registered      int64
cnt             int64
dtype: object

In [15]:
hourly_core.dtypes

instant         int64
dteday         object
season          int64
yr              int64
mnth            int64
hr              int64
holiday         int64
weekday         int64
workingday      int64
weathersit      int64
temp          float64
atemp         float64
hum           float64
windspeed     float64
casual          int64
registered      int64
cnt             int64
dtype: object

In [16]:
daily_core.duplicated().sum()

0

In [17]:
hourly_core.duplicated().sum()

0

In [18]:
daily_core.isna().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

In [19]:
hourly_core.isna().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

Apparently there are no null values, but there **are** values missing / non-existant.

Can I impute the missing values?<br />
For the missing data in *hourly_core*. The missing hourly entries could be filled. Date & hour can be sequenced, temperature and windspeed filled in from the weather site the UCI researchers used (which by the way takes its data from the weather station at Reagan National Airport located in the middle of Washington D.C. and also supplies NOAA).<br />
The number of bike riders though cannot be imputed because at least one range of missing data covers an extreme weather  event (there is only one entry for 2012-10-29 [Hurricane Sandy]).

In [20]:
# How much missing data is there in the hourly and daily data set?
print("Number of hourly entries needed:",((4*30+7*31)*2+28+29)*24, 
        "Number of hourly entries missing:", ((4*30+7*31)*2+28+29)*24 - hourly_core.instant.count())
print("Number of daily entries needed:",((4*30+7*31)*2+28+29), 
        "Number of daily entries missing:", ((4*30+7*31)*2+28+29) - daily_core.instant.count())


Number of hourly entries needed: 17544 Number of hourly entries missing: 165
Number of daily entries needed: 731 Number of daily entries missing: 0


<hr>

Little bit of data maintenance.

In [21]:
# Rename columns
hourly_core.columns = ['instant', 'date_day', 'season', 'year', 'month', 'hour', 'is_holiday', 'weekday', 'is_workday', 'weather', 'temp', 'temp_feel', 'humidity', 'windspeed', 'casual', 'registered', 'total_riders']
daily_core.columns = ['instant', 'date_day', 'season', 'year', 'month', 'is_holiday', 'weekday', 'is_workday', 'weather', 'temp', 'temp_feel', 'humidity', 'windspeed', 'casual', 'registered', 'total_riders']


In [22]:
# Convert certain numerical columns to categories with more human-friendly content
# note: the mapping of seasons is wrong both in the readme file (winter=1) and on the data set website (spring=1), 
# seasons were remapped according to WMO conventions (see note).

hourly_core['season'] = hourly_core.month.map({
        1:'winter', 2:'winter', 
        3:'spring', 4:'spring', 5:'spring', 
        6:'summer', 7:'summer', 8:'summer', 
        9:'autumn', 10:'autumn', 11:'autumn', 
        12:'winter'
    }).astype('category')
hourly_core['weekday'] = hourly_core.weekday.map({
        1:'monday', 2:'tuesday', 3:'wednesday', 4:'thursday', 5:'friday', 6:'saturday', 0:'sunday'
    }).astype('category')
hourly_core['is_holiday'] = hourly_core.is_holiday.map({0:'no', 1:'yes'}).astype('category')
hourly_core['weather_desc'] = hourly_core.weather.map({1:'clear', 2:'foggy', 3:'inclement', 4:'crap'}).astype('category')
hourly_core['year'] = hourly_core.year.map({0:2011, 1:2012}).astype('category')

daily_core['season'] = daily_core.month.map({
        1:'winter', 2:'winter', 
        3:'spring', 4:'spring', 5:'spring', 
        6:'summer', 7:'summer', 8:'summer', 
        9:'autumn', 10:'autumn', 11:'autumn', 
        12:'winter'}).astype('category')
daily_core['weekday'] = daily_core.weekday.map({
        1:'monday', 2:'tuesday', 3:'wednesday', 4:'thursday', 5:'friday', 6:'saturday', 0:'sunday'
    }).astype('category')
daily_core['is_holiday'] = daily_core.is_holiday.map({0:'no', 1:'yes'}).astype('category')
daily_core['weather_desc'] = daily_core.weather.map({1:'clear', 2:'foggy', 3:'inclement', 4:'crap'}).astype('category')
daily_core['year'] = daily_core.year.map({0:2011, 1:2012}).astype('category')


In [23]:
# Add a time of the day column (morning (06-12), afternoon (12-18), evening (18-00), night (00-06))
hourly_core['daytime'] = hourly_core.hour.apply(lambda value: "night" 
                                            if value <= 6 else "morning"
                                            if value <= 12 else "afternoon"
                                            if value <= 18 else "evening").astype('category')


In [24]:
# Convert time and date to timestamps
hourly_core['hour'] = pd.to_datetime(hourly_core['hour'], format='%H').dt.time
hourly_core['date_day'] = pd.to_datetime(hourly_core['date_day']).dt.date

daily_core['date_day'] = pd.to_datetime(daily_core['date_day']).dt.date

# Combine time and date into an extra column
hourly_core['date_time'] = hourly_core.apply(lambda r : dt.datetime.combine(r['date_day'],r['hour']),1)
daily_core['date_time'] = daily_core['date_day']


In [25]:
daily_core.dtypes

instant            int64
date_day          object
season          category
year            category
month              int64
is_holiday      category
weekday         category
is_workday         int64
weather            int64
temp             float64
temp_feel        float64
humidity         float64
windspeed        float64
casual             int64
registered         int64
total_riders       int64
weather_desc    category
date_time         object
dtype: object

In [26]:
hourly_core.dtypes

instant                  int64
date_day                object
season                category
year                  category
month                    int64
hour                    object
is_holiday            category
weekday               category
is_workday               int64
weather                  int64
temp                   float64
temp_feel              float64
humidity               float64
windspeed              float64
casual                   int64
registered               int64
total_riders             int64
weather_desc          category
daytime               category
date_time       datetime64[ns]
dtype: object

Alternative method to find out how many records are missing, saves you from counting.

In [27]:
# And set the datetime column as index
hourly_core = hourly_core.set_index(['date_time'])
daily_core = daily_core.set_index(['date_time'])

In [28]:
temp = (pd.date_range(start="2011-01-01", end="2012-12-31", freq='1H').difference(hourly_core.index)).to_frame(index=False, name='hr_missing')
print("Number of hourly records missing:", temp.count())

temp = (pd.date_range(start="2011-01-01", end="2012-12-31", freq='1d').difference(daily_core.index)).to_frame(index=False, name='day_missing')
print("Number of daily records missing:", temp.count())

Number of hourly records missing: hr_missing    165
dtype: int64
Number of daily records missing: day_missing    0
dtype: int64


We could drop some unnecessary columns:
* date_time combines yr, mnth, day, hr
* workingday == NOT is_holiday OR saturday/sunday
* instant => date_time index takes over

but they are not really bothering us, and we may be able to use the split date more easily.<br />
Maybe drop them later, if performance is an issue.

df = df.drop(columns=['col1', 'col2'])

In [29]:
# De-normalise the values for temperature, apparent temperature, humidity, 
# and windspeed
# x = x_norm * (max_y - min_y) + min_y
hourly_core['temp'] = hourly_core.temp.apply(lambda x: ((x * (39 - -8)) + -8))
hourly_core['temp_feel'] = hourly_core.temp_feel.apply(lambda x: ((x * (50 - -16)) + -16))
hourly_core['humidity'] = hourly_core.humidity.apply(lambda x: (x * 100))
hourly_core['windspeed'] = hourly_core.windspeed.apply(lambda x: (x * 67))

daily_core['temp'] = daily_core.temp.apply(lambda x: ((x * (39 - -8)) + -8)) # (x*47-8)
daily_core['temp_feel'] = daily_core.temp_feel.apply(lambda x: ((x * (50 - -16)) + -16))
daily_core['humidity'] = daily_core.humidity.apply(lambda x: x * 100)
daily_core['windspeed'] = daily_core.windspeed.apply(lambda x: x * 67)


In [30]:
hourly_core.dtypes

instant            int64
date_day          object
season          category
year            category
month              int64
hour              object
is_holiday      category
weekday         category
is_workday         int64
weather            int64
temp             float64
temp_feel        float64
humidity         float64
windspeed        float64
casual             int64
registered         int64
total_riders       int64
weather_desc    category
daytime         category
dtype: object

Note that date_day and hour are still object types and not datetime even though they were converted to timestamps.
Why?

In [31]:
# Dataframe of all missing hourly values
# Datetime as index
missing_hours = pd.date_range(start="2011-01-01", end="2012-12-31", freq='1H').difference(hourly_core.index)
missing_hours = missing_hours.to_frame(index=False, name='hr_missing')

missing_hours['date'] = pd.to_datetime(missing_hours['hr_missing']).dt.date
missing_hours['time'] = pd.to_datetime(missing_hours['hr_missing']).dt.time

missing_hours = missing_hours.set_index(['hr_missing'])

In [32]:
missing_hours.head()

date      time
hr_missing                               
2011-01-02 05:00:00  2011-01-02  05:00:00
2011-01-03 02:00:00  2011-01-03  02:00:00
2011-01-03 03:00:00  2011-01-03  03:00:00
2011-01-04 03:00:00  2011-01-04  03:00:00
2011-01-05 03:00:00  2011-01-05  03:00:00

In [33]:
# Get the records of a specific date
filtered_df = hourly_core.loc['2011-01-02']
filtered_df

instant    date_day  season  year  month      hour  \
date_time                                                                 
2011-01-02 00:00:00       25  2011-01-02  winter  2011      1  00:00:00   
2011-01-02 01:00:00       26  2011-01-02  winter  2011      1  01:00:00   
2011-01-02 02:00:00       27  2011-01-02  winter  2011      1  02:00:00   
2011-01-02 03:00:00       28  2011-01-02  winter  2011      1  03:00:00   
2011-01-02 04:00:00       29  2011-01-02  winter  2011      1  04:00:00   
2011-01-02 06:00:00       30  2011-01-02  winter  2011      1  06:00:00   
2011-01-02 07:00:00       31  2011-01-02  winter  2011      1  07:00:00   
2011-01-02 08:00:00       32  2011-01-02  winter  2011      1  08:00:00   
2011-01-02 09:00:00       33  2011-01-02  winter  2011      1  09:00:00   
2011-01-02 10:00:00       34  2011-01-02  winter  2011      1  10:00:00   
2011-01-02 11:00:00       35  2011-01-02  winter  2011      1  11:00:00   
2011-01-02 12:00:00       36  2011-01-02  winter  2011      1  12:00:00   
2011-01-02 13:00:00       37  2011-01-02  winter  2011      1  13:00:00   
2011-01-02 14:00:00       38  2011-01-02  winter  2011      1  14:00:00   
2011-01-02 15:00:00       39  2011-01-02  winter  2011      1  15:00:00   
2011-01-02 16:00:00       40  2011-01-02  winter  2011      1  16:00:00   
2011-01-02 17:00:00       41  2011-01-02  winter  2011      1  17:00:00   
2011-01-02 18:00:00       42  2011-01-02  winter  2011      1  18:00:00   
2011-01-02 19:00:00       43  2011-01-02  winter  2011      1  19:00:00   
2011-01-02 20:00:00       44  2011-01-02  winter  2011      1  20:00:00   
2011-01-02 21:00:00       45  2011-01-02  winter  2011      1  21:00:00   
2011-01-02 22:00:00       46  2011-01-02  winter  2011      1  22:00:00   
2011-01-02 23:00:00       47  2011-01-02  winter  2011      1  23:00:00   

                    is_holiday weekday  is_workday  weather   temp  temp_feel  \
date_time                                                                       
2011-01-02 00:00:00         no  sunday           0        2  13.62    13.9970   
2011-01-02 01:00:00         no  sunday           0        2  12.68    13.0004   
2011-01-02 02:00:00         no  sunday           0        2  11.74    11.9972   
2011-01-02 03:00:00         no  sunday           0        2  13.62    13.9970   
2011-01-02 04:00:00         no  sunday           0        2  13.62    13.9970   
2011-01-02 06:00:00         no  sunday           0        3  11.74    11.9972   
2011-01-02 07:00:00         no  sunday           0        2  10.80    11.0006   
2011-01-02 08:00:00         no  sunday           0        3  10.80    11.0006   
2011-01-02 09:00:00         no  sunday           0        2   9.86     9.9974   
2011-01-02 10:00:00         no  sunday           0        2   8.92     7.0010   
2011-01-02 11:00:00         no  sunday           0        2   8.92     5.9978   
2011-01-02 12:00:00         no  sunday           0        2   8.92     5.9978   
2011-01-02 13:00:00         no  sunday           0        2   8.92     7.0010   
2011-01-02 14:00:00         no  sunday           0        3   8.92     7.0010   
2011-01-02 15:00:00         no  sunday           0        3   7.98     5.9978   
2011-01-02 16:00:00         no  sunday           0        3   7.98     5.9978   
2011-01-02 17:00:00         no  sunday           0        1   7.98     5.9978   
2011-01-02 18:00:00         no  sunday           0        2   8.92     5.9978   
2011-01-02 19:00:00         no  sunday           0        1   7.04     3.0014   
2011-01-02 20:00:00         no  sunday           0        1   6.10     1.9982   
2011-01-02 21:00:00         no  sunday           0        1   4.22    -0.9982   
2011-01-02 22:00:00         no  sunday           0        1   3.28    -2.0014   
2011-01-02 23:00:00         no  sunday           0        1   2.34    -0.9982   

                     humidity  windspeed  casual  registered  total_riders  \
date_time                                

In [34]:
# Get the records for Hurricane Sandy
# Oct 26: state of emergency declared in the US
# Oct 29-30: all government buildings closed and DC Metro services suspended
filtered_df = hourly_core['2012-10-26':'2012-10-30']
filtered_df.weather.unique()

# For this date range the weather is clear (1), misty (2), and light rain (3).
# Do we agree that a hurricane is at least a 4? Heavy rain and the like?


array([2, 1, 3])

Taking as an example a day where there is only one missing value. If the number of riders (casual, registered, total) is different to the sum of all 23 values present in the hourly data of the day then we can fill in the rider data with the difference and take the weather information from the Washington National Airport weather station.

Result: Checked several days, the difference was always zero. It is unlikely that they all had 0 riders at that time. The sums of the riders in the hourly data from the example day matches the recorded riders from that day's daily record, therefore the daily records were computed using the hourly data with missing records not substituted in any way.

In [35]:
# Example day
filtered_df = hourly_core.loc['2011-01-04']

# Values for that day (2011-01-02): casual: 131, registered: 670, total: 801
# Values for (2011-01-04): 108	1454	1562
print("casuals:", filtered_df.casual.sum())
print("registered:", filtered_df.registered.sum())

casuals: 108
registered: 1454


In [36]:
# Some things to make categories of: cold, hot, and muggy hours and days
# NOAA typically considers relative humidity (RH) levels of 50% or more, and dewpoints 
# (a more direct measure of humidity) above 65 F (18 C) to be uncomfortably high

hourly_core['cold'] = np.where(hourly_core.temp_feel < 5, 1, 0)
hourly_core['hot'] = np.where(hourly_core.temp_feel > 30, 1, 0)
hourly_core['muggy'] = np.where(hourly_core.humidity >= 50, 1, 0)

daily_core['cold'] = np.where(daily_core.temp_feel < 5, 1, 0)
daily_core['hot'] = np.where(daily_core.temp_feel > 30, 1, 0)
daily_core['muggy'] = np.where(daily_core.humidity >= 50, 1, 0)

In [37]:
hourly_core.reset_index(inplace=True)
hourly_core.to_csv('data/mod_data/hour_mod.csv', index=False)
daily_core.reset_index(inplace=True)
daily_core.to_csv('data/mod_data/day_mod.csv', index=False)

<hr>

Adding the trip history data for 2011 and 2012.

In [38]:
trips_2011 = pd.read_csv('data/cbs_data/2011-capitalbikeshare-tripdata.csv')

path = r'data/cbs_data/2012'
all_files = glob.glob(path + "/*.csv")

li = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=None, header=0)
    li.append(df)

trips_2012 = pd.concat(li, axis=0, ignore_index=True)

In [39]:
trips_2011.columns = trips_2011.columns.str.replace(' ','_').str.lower()
trips_2012.columns = trips_2012.columns.str.replace(' ','_').str.lower()


In [47]:
trips_2011.head()

duration           start_date             end_date  start_station_number  \
0      3548  2011-01-01 00:01:29  2011-01-01 01:00:37                 31620   
1       346  2011-01-01 00:02:46  2011-01-01 00:08:32                 31105   
2       562  2011-01-01 00:06:13  2011-01-01 00:15:36                 31400   
3       434  2011-01-01 00:09:21  2011-01-01 00:16:36                 31111   
4       233  2011-01-01 00:28:26  2011-01-01 00:32:19                 31104   

                    start_station  end_station_number  \
0                   5th & F St NW               31620   
1            14th & Harvard St NW               31101   
2  Georgia & New Hampshire Ave NW               31104   
3                  10th & U St NW               31503   
4     Adams Mill & Columbia Rd NW               31106   

                   end_station bike_number member_type  
0                5th & F St NW      W00247      Member  
1               14th & V St NW      W00675      Casual  
2  Adams Mill & Columbia Rd NW      W00357      Member  
3        Florida Ave & R St NW      W00970      Member  
4     Calvert & Biltmore St NW      W00346      Casual

In [48]:
trips_2012.head()

duration           start_date             end_date  start_station_number  \
0       192  2012-04-01 00:01:13  2012-04-01 00:04:26                 31237   
1       978  2012-04-01 00:01:49  2012-04-01 00:18:08                 31225   
2       843  2012-04-01 00:03:25  2012-04-01 00:17:28                 31230   
3       408  2012-04-01 00:03:37  2012-04-01 00:10:26                 31110   
4       163  2012-04-01 00:04:13  2012-04-01 00:06:57                 31104   

                    start_station  end_station_number  \
0   25th St & Pennsylvania Ave NW               31212   
1  C & O Canal & Wisconsin Ave NW               31228   
2   Metro Center / 12th & G St NW               31201   
3        20th St & Florida Ave NW               31239   
4     Adams Mill & Columbia Rd NW               31112   

                         end_station bike_number member_type  
0                     21st & M St NW      W00663      Member  
1                      8th & H St NW      W00574      Member  
2                     15th & P St NW      W00201      Member  
3  Rhode Island & Connecticut Ave NW      W01060      Casual  
4      Harvard St & Adams Mill Rd NW      W01071      Casual

In [56]:
# Add field to each data frame
trips_2011['year'] = '2011'
trips_2012['year'] = '2012'

# and concatenate
bike_trips = pd.concat([trips_2011, trips_2012],axis=0)

# Give some more detail
bike_trips.rename(columns={'duration': 'duration_sec'}, inplace=True)


In [57]:
bike_trips.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3255678 entries, 0 to 2028910
Data columns (total 10 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   duration_sec          int64 
 1   start_date            object
 2   end_date              object
 3   start_station_number  int64 
 4   start_station         object
 5   end_station_number    int64 
 6   end_station           object
 7   bike_number           object
 8   member_type           object
 9   year                  object
dtypes: int64(3), object(7)
memory usage: 273.2+ MB


In [58]:
# Some maintenance
bike_trips = bike_trips.astype({'start_date': 'datetime64', 'end_date': 'datetime64', 'member_type': 'category', 'year': 'category'})

bike_trips['date'] = bike_trips['start_date'].dt.date
bike_trips['hour'] = bike_trips['start_date'].dt.hour

# Convert time and date to timestamps
bike_trips['date'] = pd.to_datetime(bike_trips['date']).dt.date
bike_trips['hour'] = pd.to_datetime(bike_trips['hour'], format='%H').dt.time

# Combine time and date into an extra column
bike_trips['date_hour'] = bike_trips.apply(lambda r : dt.datetime.combine(r['date'],r['hour']),1)

# We could drop these but I'm hanging on to them for now
#bike_trips.drop(['date', 'hour'], axis=1, inplace=True)

In [59]:
bike_trips.nunique()

duration_sec              20865
start_date              3092552
end_date                3083001
start_station_number        194
start_station               194
end_station_number          194
end_station                 194
bike_number                1771
member_type                   3
year                          2
date                        731
hour                         24
date_hour                 17379
dtype: int64

There are 194 stations and 1771 bikes.

In [65]:
# There are 21 records where the member type is 'Unknown'
bike_trips.loc[bike_trips.member_type == 'Unknown']

duration_sec          start_date            end_date  \
414633            501 2011-06-09 14:34:39 2011-06-09 14:43:00   
424868           2233 2011-06-11 17:40:02 2011-06-11 18:17:15   
787755            257 2011-08-31 10:17:31 2011-08-31 10:21:48   
1059539           443 2011-11-06 01:29:12 2011-11-06 01:36:35   
112372           1341 2012-04-19 16:23:59 2012-04-19 16:46:20   
114829           1039 2012-04-19 20:17:07 2012-04-19 20:34:27   
118863            384 2012-04-20 14:34:16 2012-04-20 14:40:41   
495782           2242 2012-06-20 00:05:36 2012-06-20 00:42:59   
576117            959 2012-10-02 17:59:28 2012-10-02 18:15:27   
693772            577 2012-10-20 03:00:16 2012-10-20 03:09:54   
785039            478 2012-11-04 19:04:49 2012-11-04 19:12:47   
785631            891 2012-11-05 01:06:36 2012-11-05 01:21:27   
785639            305 2012-11-05 02:06:15 2012-11-05 02:11:21   
819296            893 2012-11-11 01:18:43 2012-11-11 01:33:36   
920317            982 2012-12-02 01:03:36 2012-12-02 01:19:59   
1195504          1181 2012-02-18 17:20:47 2012-02-18 17:40:29   
1365214           125 2012-03-26 01:47:39 2012-03-26 01:49:45   
1479678           745 2012-07-13 19:40:01 2012-07-13 19:52:27   
1599367          1671 2012-07-31 20:02:31 2012-07-31 20:30:22   
1599571           242 2012-07-31 20:27:17 2012-07-31 20:31:19   
1599602           499 2012-07-31 20:30:42 2012-07-31 20:39:01   

         start_station_number                              start_station  \
414633                  31205                             21st & I St NW   
424868                  31215             Georgetown Harbor / 30th St NW   
787755                  31218             L'Enfant Plaza / 7th & C St SW   
1059539                 31010                      S Glebe & Potomac Ave   
112372                  31014                       Lynn & 19th St North   
114829                  31215             Georgetown Harbor / 30th St NW   
118863                  31234              20th & O St NW / Dupont South   
495782                  31235              19th St & Constitution Ave NW   
576117                  31230              Metro Center / 12th & G St NW   
693772                  31102                        11th & Kenyon St NW   
785039                  31200       Massachusetts Ave & Dupont Circle NW   
785631                  31241                              Thomas Circle   
785639                  31602                    Park Rd & Holmead Pl NW   
819296                  31228                              8th & H St NW   
920317                  31257              22nd & I St NW / Foggy Bottom   
1195504                 31232  7th & F St NW / National Portrait Gallery   
1365214                 31241                              Thomas Circle   
1479678                 31232  7th & F St NW / National Portrait Gallery   
1599367                 31231    14th & D St NW / Ronald Reagan Building   
1599571                 31234              20th & O St NW / Dupont South   
1599602                 31623            Columbus Circle / Union Station   

         end_station_number                                   end_station  \
414633                31232     7th & F St NW / National Portrait Gallery   
424868                31214                         17th & Corcoran St NW   
787755                31219                 10th St & Constitution Ave NW   
1059539               31013                              Eads & 22nd St S   
112372                31007           Crystal City Metro / 18th & Bell St   
114829                31609                         Maine Ave & 7th St SW   
118863                31246                    M St & Pennsylvania Ave NW   
495782                31220  US Dept of State / Virginia Ave & 21st St NW   
576117                31603                                 1st & M St NE   
693772                31203                    14th & Rhode Island Ave NW   
785039                31202                               

In [67]:
# Check the 21 records where the member_type is 'Unknown' and see if they are part of the hourly records.

hourly_core[hourly_core.date_time == '2012-11-05 02:00:00'].\
            groupby(['casual', 'registered', 'total_riders']).count()

date_time  instant  date_day  season  year  \
casual registered total_riders                                               
2      4          6                     1        1         1       1     1   

                                month  hour  is_holiday  weekday  is_workday  \
casual registered total_riders                                                 
2      4          6                 1     1           1        1           1   

                                weather  temp  temp_feel  humidity  windspeed  \
casual registered total_riders                                                  
2      4          6                   1     1          1         1          1   

                                weather_desc  daytime  cold  hot  muggy  
casual registered total_riders                                           
2      4          6                        1        1     1    1      1

In [68]:
bike_trips[bike_trips.date_hour == '2012-11-05 02:00:00'].groupby(['member_type']).count()

# Result: The hourly records disregard the instances of the trip data sets where the member_type is 'Unknown'
# Should probably drop these instances but will keep them in case I map bike routes

duration_sec  start_date  end_date  start_station_number  \
member_type                                                             
Casual                  2           2         2                     2   
Member                  4           4         4                     4   
Unknown                 1           1         1                     1   

             start_station  end_station_number  end_station  bike_number  \
member_type                                                                
Casual                   2                   2            2            2   
Member                   4                   4            4            4   
Unknown                  1                   1            1            1   

             year  date  hour  date_hour  
member_type                               
Casual          2     2     2          2  
Member          4     4     4          4  
Unknown         1     1     1          1

In [69]:
bike_trips.head()

duration_sec          start_date            end_date  start_station_number  \
0          3548 2011-01-01 00:01:29 2011-01-01 01:00:37                 31620   
1           346 2011-01-01 00:02:46 2011-01-01 00:08:32                 31105   
2           562 2011-01-01 00:06:13 2011-01-01 00:15:36                 31400   
3           434 2011-01-01 00:09:21 2011-01-01 00:16:36                 31111   
4           233 2011-01-01 00:28:26 2011-01-01 00:32:19                 31104   

                    start_station  end_station_number  \
0                   5th & F St NW               31620   
1            14th & Harvard St NW               31101   
2  Georgia & New Hampshire Ave NW               31104   
3                  10th & U St NW               31503   
4     Adams Mill & Columbia Rd NW               31106   

                   end_station bike_number member_type  year        date  \
0                5th & F St NW      W00247      Member  2011  2011-01-01   
1               14th & V St NW      W00675      Casual  2011  2011-01-01   
2  Adams Mill & Columbia Rd NW      W00357      Member  2011  2011-01-01   
3        Florida Ave & R St NW      W00970      Member  2011  2011-01-01   
4     Calvert & Biltmore St NW      W00346      Casual  2011  2011-01-01   

       hour  date_hour  
0  00:00:00 2011-01-01  
1  00:00:00 2011-01-01  
2  00:00:00 2011-01-01  
3  00:00:00 2011-01-01  
4  00:00:00 2011-01-01

In [70]:
duplicates = bike_trips[bike_trips.duplicated()]
duplicates

Empty DataFrame
Columns: [duration_sec, start_date, end_date, start_station_number, start_station, end_station_number, end_station, bike_number, member_type, year, date, hour, date_hour]
Index: []

Nice! No duplicates.

In [71]:
bike_trips.describe().T

count          mean          std      min      25%  \
duration_sec          3255678.0   1084.497036  2100.266770     60.0    399.0   
start_station_number  3255678.0  31289.742469   193.121189  31000.0  31200.0   
end_station_number    3255678.0  31291.821241   192.376122  31000.0  31201.0   

                          50%      75%      max  
duration_sec            658.0   1092.0  86355.0  
start_station_number  31226.0  31401.0  31806.0  
end_station_number    31225.0  31402.0  31806.0

There is a minimum of 60 seconds for a bike trip. That is a bit short.<br />
Also a maximum of 86355 seconds, that is 45 seconds short of 24 hours.<br />
We'll take a look at that later.

<hr>

Cleaning up the bike trips data.

In [75]:
# Some cosmetic changes. There are 19329 records where the prefix of 
# the bike number is in lowercase.

bike_trips['bike_number'] = bike_trips['bike_number'].str.upper()

In [79]:
# check and remove incomplete records: no bike number

# Check if the bike number is invalid, i.e. not W+four digits
no_bike = bike_trips[~bike_trips['bike_number'].str.contains('W')]

# There are 4801 records that do not contain valid bike numbers
# The bike number slots are filled by 16 different codes in a 
# 0X-16 digits and letters format
no_bike.bike_number.nunique()

# remove the records with no valid bike numbers
bike_trips.drop(bike_trips[~bike_trips['bike_number'].str.contains('W')].index, inplace=True)

In [80]:
# check and remove incomplete records: no member type == Unknown
no_member = bike_trips[bike_trips['member_type'] == 'Unknown']
no_member.count()

# There are 21 records where the member type is unknown. 
# These were also not entered in the daily or hourly record.
bike_trips.drop(bike_trips[bike_trips['member_type'] == 'Unknown'].index, inplace=True)

In [87]:
# Check for non standard start stations = NONE and start station numbers = NONE
#bike_trips.start_station.unique()
#bike_trips.start_station.nunique()

# Check for non standard end stations = NONE and end station numbers = NONE
bike_trips.end_station.unique()
#bike_trips.end_station.nunique()

# There are 194 unique start and end stations. No empty values.

194

In [88]:
# Check for stalled transactions or users that reconsidered: 
# same start and end station and the duration of ride is 60 seconds or less
# There are 136 records with these specifications

stalled = bike_trips.loc[(bike_trips['duration_sec'] <= 60) & (bike_trips['start_station_number'] == bike_trips['end_station_number'])]

# Remove the 60 second bike ride records
bike_trips.drop(bike_trips[(bike_trips['duration_sec'] <= 60) & (bike_trips['start_station_number'] == bike_trips['end_station_number'])].index, inplace=True)

In [89]:
# Check for bike rides that are 60 seconds or less 
# and have different start and stop stations
# There are 27 of these
bit_moved = bike_trips.loc[(bike_trips['duration_sec'] <= 60) & (bike_trips['start_station_number'] != bike_trips['end_station_number'])]

# Of these all except one lie very close together (customer changed their mind?) 
# and one (ID 507845) is 3.5 miles apart which is impossible
# to drive in 1 minute (12 minutes by car)
# Remove these 27 records

bike_trips.drop(bike_trips[(bike_trips['duration_sec'] <= 60) & (bike_trips['start_station_number'] != bike_trips['end_station_number'])].index, inplace=True)

In [90]:
# Check for bike rides that are over 23 hours. 
# Either the customer forgot to book it back in, the endpoint did not accept the 
# transaction, or something else happened. Gives a fine of $1200, so strong incentive 
# to make sure to book back in.
# There are 6 of these
long_ride = bike_trips.loc[(bike_trips['duration_sec'] >= 86_000)] 

# Remove these 6 records
bike_trips.drop(bike_trips[(bike_trips['duration_sec'] >= 86_000)].index, inplace=True)


In [103]:
bike_trips.to_csv('data/mod_data/bike_trips_2011_2012.csv')

<hr>

Supplementary material: Station information

In [97]:
# Get station information from json file

with open('data/cbs_data/station_information.json', 'r') as json_file:
	json_load = json.load(json_file)

station_info = pd.DataFrame(json_load['data']['stations'])

In [93]:
station_info.head()

name  \
0                           Eads St & 15th St S   
1                        Crystal Dr & 20th St S   
2                        Crystal Dr & 15th St S   
3  Aurora Hills Cmty Ctr / 18th St & S Hayes St   
4    Pentagon City Metro / 12th St & S Hayes St   

                            external_id        lon     rental_methods  \
0  082469cc-1f3f-11e7-bf6b-3863bb334450 -77.053230  [CREDITCARD, KEY]   
1  08246c35-1f3f-11e7-bf6b-3863bb334450 -77.049232  [CREDITCARD, KEY]   
2  08246cd5-1f3f-11e7-bf6b-3863bb334450 -77.049417  [CREDITCARD, KEY]   
3  08246d68-1f3f-11e7-bf6b-3863bb334450 -77.059490  [CREDITCARD, KEY]   
4  08246df5-1f3f-11e7-bf6b-3863bb334450 -77.059936  [CREDITCARD, KEY]   

  region_id  electric_bike_surcharge_waiver legacy_id  has_kiosk  capacity  \
0        41                           False         1       True        15   
1        41                           False         3       True        17   
2        41                           False         4       True        16   
3        41                           False         5       True        12   
4        41                           False         6       True        19   

         lat station_id station_type  eightd_has_key_dispenser  \
0  38.858971          1      classic                     False   
1  38.856425          3      classic                     False   
2  38.861056          4      classic                     False   
3  38.857866          5      classic                     False   
4  38.862303          6      classic                     False   

                                         rental_uris short_name  \
0  {'ios': 'https://dc.lft.to/lastmile_qr_scan', ...      31000   
1  {'ios': 'https://dc.lft.to/lastmile_qr_scan', ...      31002   
2  {'ios': 'https://dc.lft.to/lastmile_qr_scan', ...      31003   
3  {'ios': 'https://dc.lft.to/lastmile_qr_scan', ...      31004   
4  {'ios': 'https://dc.lft.to/lastmile_qr_scan', ...      31005   

  eightd_station_services  
0                      []  
1                      []  
2                      []  
3                      []  
4                      []

In [94]:
station_info.columns

Index(['name', 'external_id', 'lon', 'rental_methods', 'region_id',
       'electric_bike_surcharge_waiver', 'legacy_id', 'has_kiosk', 'capacity',
       'lat', 'station_id', 'station_type', 'eightd_has_key_dispenser',
       'rental_uris', 'short_name', 'eightd_station_services'],
      dtype='object')

In [98]:
# Drop not needed columns in station information

station_info = station_info.drop(['external_id', 'rental_methods', 'electric_bike_surcharge_waiver', 'has_kiosk', 'station_id',\
	'station_type', 'eightd_has_key_dispenser', 'rental_uris', 'eightd_station_services'], axis = 1)


In [100]:
station_info.rename(columns={'lon': 'longitude', 'lat': 'latitude'}, inplace=True)

In [101]:
station_info.head()

name  longitude region_id  \
0                           Eads St & 15th St S -77.053230        41   
1                        Crystal Dr & 20th St S -77.049232        41   
2                        Crystal Dr & 15th St S -77.049417        41   
3  Aurora Hills Cmty Ctr / 18th St & S Hayes St -77.059490        41   
4    Pentagon City Metro / 12th St & S Hayes St -77.059936        41   

  legacy_id  capacity   latitude short_name  
0         1        15  38.858971      31000  
1         3        17  38.856425      31002  
2         4        16  38.861056      31003  
3         5        12  38.857866      31004  
4         6        19  38.862303      31005

In [102]:
station_info.to_csv('data/mod_data/stationinfo_conv.csv')